hathor/transaction/vertex.py

```diff
@@ -0,0 +1,52 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from dataclasses import dataclass
+from typing import Generic, TypeAlias, TypeVar
+
+from hathor.transaction import BaseTransaction, Block, MergeMinedBlock, Transaction
+from hathor.transaction.token_creation_tx import TokenCreationTransaction
+
+T = TypeVar('T', bound=BaseTransaction)
+
+
+@dataclass(frozen=True, slots=True)
+class _VertexWrapper(Generic[T]):
+    """A simple generic wrapper around a BaseTransaction used to create each tag of the Vertex tagged union."""
+    base_tx: T
+
+
+class BlockType(_VertexWrapper[Block]):
+    """Vertex wrapper for Block."""
+
+
+class MergeMinedBlockType(_VertexWrapper[MergeMinedBlock]):
+    """Vertex wrapper for MergeMinedBlock."""
+
+
+class TransactionType(_VertexWrapper[Transaction]):
+    """Vertex wrapper for Transaction."""
+
+
+class TokenCreationTransactionType(_VertexWrapper[TokenCreationTransaction]):
+    """Vertex wrapper for TokenCreationTransaction."""
+
+
+"""
+A Vertex algebraic sum type, or tagged union, that unifies all vertex types.
+
+This is useful when dealing with vertices in functional code, for example when using `match` statements, so we don't
+have problems with inheritance in `case` branches.
+"""
+Vertex: TypeAlias = BlockType | MergeMinedBlockType | TransactionType | TokenCreationTransactionType
```

Comment:
Why are we doing this? I think it is confusing to have two different types for each vertex type.

Reply:
Indeed this is not strictly necessary. It's an improvement, however, in my opinion. This uses the [tagged unions](https://en.wikipedia.org/wiki/Tagged_union) concept of type theory, and I believe we could benefit from using this in future code. This is available in other forms in many languages (like sealed interfaces in Kotlin, or Enums in Rust), and is the same concept behind `Result` and `Option` types (that I also believe we should start using).

They're simply a low-cost wrapper for the existing `BaseTransaction` subclasses, with the advantage that they represent a single type. If we have a `Block`, for example, we have no way to know if it is a `Block` or a `MergeMinedBlock`, unless we perform a check. With this new `BlockType`, we're sure this is strictly a `Block`, and not any subclass (well, restricted via semantics, but we could restrict this via code too). This is useful for verification, as we have different behavior for each subclass, and any other code that perform this kind of polymorphism.

I do concede it may be a bit confusing at this time, however I think the types are not exactly redundant as they provide different semantics. And I also think if we end up using them in other code, we may find we want to evolve the current models in the future, and maybe this would become the main form of abstraction—by using a data class for each vertex type, after we remove all dependencies from the existing classes (verification, storage, etc). So I like the trade-off.

I could certainly remove it now, if you and @jansegre disagree. But at worst I see this as an experiment that we could remove down the road, if we don't like it.

Specifically for the use case in this PR, it removed boilerplate, as it allowed to change this:

```python
# We assert with type() instead of isinstance() because each subclass has a specific branch.
match vertex.version:
    case TxVersion.REGULAR_BLOCK:
        assert type(vertex) is Block
        self._verify_without_storage_block(vertex)
    case TxVersion.MERGE_MINED_BLOCK:
        assert type(vertex) is MergeMinedBlock
        self._verify_without_storage_merge_mined_block(vertex)
    case TxVersion.REGULAR_TRANSACTION:
        assert type(vertex) is Transaction
        self._verify_without_storage_tx(vertex)
    case TxVersion.TOKEN_CREATION_TRANSACTION:
        assert type(vertex) is TokenCreationTransaction
        self._verify_without_storage_token_creation_tx(vertex)
    case _:
        assert_never(vertex.version)
```

To this:

```python
vertex = base_tx.as_vertex()
match vertex:
    case BlockType(block):
        self._verify_without_storage_block(block)
    case MergeMinedBlockType(block):
        self._verify_without_storage_merge_mined_block(block)
    case TransactionType(tx):
        self._verify_without_storage_tx(tx)
    case TokenCreationTransactionType(tx):
        self._verify_without_storage_token_creation_tx(tx)
    case _:
        assert_never(vertex)
```

In several places.

Reply:
I've removed this from this PR.